### PR TITLE
Add workspace invariant of valid baseURL

### PR DESCRIPTION
**Pull request by sttts**
This PR does the following:
1. To simplify the state machine as seen by users and controllers, this PR adds the invariant that

   - After scheduling `status.baseURL` of a ClusterWorkspace stays valid.

   In other words: there is no unscheduling. A workspace has state (obviously) and a workspace without a shard does not make sense.

   Corrolary: We have to make sure (through movement and high availability measures) that a workspace is available. Clients should not be expected to hande the case that a workspace is not accessible. Reason:
   - client-go clients don't watch the workspace object.
   - processes talk to some URL via clients, and clients cannot update the URL they talk to (that's not part of the Kubernetes client contract).

   tl/dr: we have to make sure the workspace baseURL keeps working at all time.
2. A `ClusterWorkspace` admission plugin is added verifying the state transitions of cluster workspace objects.

### CODE REVIEW

File: pkg/reconciler/workspace/controller.go [hunk 1]
```diff
@@ -286,39 +286,67 @@ func (c *Controller) process(ctx context.Context, key string) error {
...
+		if current := workspace.Status.Location.Current; current != "" {
+			// make sure current shard still exists
+			if _, err := c.rootWorkspaceShardLister.Get(clusters.ToClusterAwareKey(tenancyhelper.RootCluster, current)); errors.IsNotFound(err) {
+				klog.Infof("De-scheduling workspace %q|%q from nonexistent shard %q", tenancyhelper.RootCluster, workspace.Name, current)
+				workspace.Status.Location.Current = ""
+			} else if err != nil {
+				return err
+			}
 		}
-		shard = currentShard
```
Comment (stevekuznetsov) [inline pkg/reconciler/workspace/controller.go:299]: Why did we lose this?

File: test/e2e/reconciler/workspace/controller_test.go [hunk 2]
```diff
@@ -331,9 +340,11 @@ func unschedulable(object *tenancyv1alpha1.ClusterWorkspace) error {
 func scheduled(target string) func(workspace *tenancyv1alpha1.ClusterWorkspace) error {
 	return func(object *tenancyv1alpha1.ClusterWorkspace) error {
 		if isUnschedulable(object) {
+			klog.Infof("workspace %s is unschedulable, skipping scheduled check", object.Name)
```
Comment (stevekuznetsov) [inline test/e2e/reconciler/workspace/controller_test.go:343]: Why do we need the klogs, should the errors not be logged by `t.Error`?

File: pkg/reconciler/workspace/controller.go (hunk 1 above)
Comment (stevekuznetsov) [reply pkg/reconciler/workspace/controller.go:299]: Flow seems correct but is a bit more implicit this way.

**Review by stevekuznetsov — APPROVED** [review]
Seems fine to me - how common is the admission pattern?

File: pkg/reconciler/workspace/controller.go (hunk 1 above)
Comment (sttts) [reply pkg/reconciler/workspace/controller.go:299]: are you referring to line 375 where we reget the shard?

File: test/e2e/reconciler/workspace/controller_test.go (hunk 2 above)
Comment (sttts) [reply test/e2e/reconciler/workspace/controller_test.go:343]: Just to have some indication when things go wrong. Silent wait loops are usually silent despite the logs when things are good, but should be noisy when things are bad.

File: pkg/reconciler/workspace/controller.go (hunk 1 above)
Comment (sttts) [reply pkg/reconciler/workspace/controller.go:299]: updated. Actually, we only want to touch the baseURL during scheduling. After that it is set in stone forever.

File: pkg/admission/clusterworkspace/admission.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		if err != nil {
+			return fmt.Errorf("unexpected unknown old object, got %v, expected ClusterWorkspace", a.GetOldObject().GetObjectKind().GroupVersionKind().Kind)
+		}
+		old, ok := obj.(*tenancyv1alpha1.ClusterWorkspace)
+		if !ok {
+			return fmt.Errorf("unexpected unknown old object, got %v, expected ClusterWorkspace", obj.GetObjectKind().GroupVersionKind().Kind)
+		}
+
+		if old.Spec.Type != cw.Spec.Type {
+			return admission.NewForbidden(a, errors.New("spec.type is immutable"))
```
Comment (ncdc) [inline pkg/admission/clusterworkspace/admission.go:94]: Do you want to use any validation helpers, such as ValidateImmutableField?

File: pkg/admission/clusterworkspace/admission_test.go [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+						Phase:        tenancyv1alpha1.ClusterWorkspacePhaseReady,
+						Initializers: []tenancyv1alpha1.ClusterWorkspaceInitializer{},
+						Location:     tenancyv1alpha1.ClusterWorkspaceLocation{Current: "somewhere"},
+						BaseURL:      "https://kcp.bigcorp.com/clusters/org:test",
+					},
+				}),
+			wantErr: true,
+		},
+		{
+			name: "ignors different resources",
```
Comment (ncdc) [inline pkg/admission/clusterworkspace/admission_test.go:309]:
```suggestion
			name: "ignores different resources",
```

File: pkg/admission/clusterworkspacetypeexists/admission_test.go [hunk 5]
```diff
@@ -454,86 +427,8 @@ func TestValidate(t *testing.T) {
...
-						Phase:        tenancyv1alpha1.ClusterWorkspacePhaseInitializing,
-						Initializers: []tenancyv1alpha1.ClusterWorkspaceInitializer{"a"},
-						Location:     tenancyv1alpha1.ClusterWorkspaceLocation{Current: "somewhere"},
-						BaseURL:      "https://kcp.bigcorp.com/clusters/org:test",
-					},
-				}),
-		},
-		{
-			name: "ignors different resources",
+			name:  "ignors different resources",
```
Comment (ncdc) [inline pkg/admission/clusterworkspacetypeexists/admission_test.go:430]:
```suggestion
			name:  "ignores different resources",
```

File: pkg/reconciler/workspace/controller.go [hunk 6]
```diff
@@ -286,39 +288,105 @@ func (c *Controller) process(ctx context.Context, key string) error {
...
+				}
+				u.Path = path.Join(u.Path, targetShard.Status.ConnectionInfo.APIPath, "clusters", logicalCluster)
+
+				workspace.Status.BaseURL = u.String()
+				workspace.Status.Location.Current = targetShard.Name
+
+				conditions.MarkTrue(workspace, tenancyv1alpha1.WorkspaceScheduled)
+				klog.Infof("Scheduled workspace %q|%q to %q|%q", workspace.ClusterName, workspace.Name, targetShard.ClusterName, targetShard.Name)
+			} else {
+				conditions.MarkFalse(workspace, tenancyv1alpha1.WorkspaceScheduled, tenancyv1alpha1.WorkspaceReasonUnschedulable, conditionsv1alpha1.ConditionSeverityError, "No available shards to schedule workspace %q|%q.", workspace.ClusterName, workspace.Name)
```
Comment (ncdc) [inline pkg/reconciler/workspace/controller.go:356]: Since this is a condition on the workspace itself, can we omit the ClusterName and Name from the message?

File: pkg/reconciler/workspace/controller.go [hunk 7]
```diff
@@ -286,39 +288,105 @@ func (c *Controller) process(ctx context.Context, key string) error {
...
+				failures := make([]string, 0, len(invalidShards))
+				for name, x := range invalidShards {
+					failures = append(failures, fmt.Sprintf("  %s: reason %q, message %q", name, x.reason, x.message))
+				}
+				klog.Infof("No valid shards found for workspace %q|%q, skipped:\n%s", workspace.ClusterName, workspace.Name, strings.Join(failures, "\n"))
```
Comment (ncdc) [inline pkg/reconciler/workspace/controller.go:361]: Ditto - can we omit workspace clusterName and name from the message?

File: pkg/admission/clusterworkspace/admission.go (hunk 3 above)
Comment (sttts) [reply pkg/admission/clusterworkspace/admission.go:94]: oh, didn't know those exists

File: pkg/reconciler/workspace/controller.go (hunk 7 above)
Comment (sttts) [reply pkg/reconciler/workspace/controller.go:361]: this is for logging, not exposed

**Comment by sttts** [comment]
@ncdc Addressed comments.

File: test/e2e/virtual/workspaces/virtual_workspace_test.go [hunk 8]
```diff
@@ -182,8 +182,8 @@ func TestWorkspacesVirtualWorkspaces(t *testing.T) {
 							return false, nil
 						}
 						return false, err
-					} else if !conditions.IsTrue(cw, tenancyv1alpha1.WorkspaceURLValid) {
-						return false, fmt.Errorf("ClusterWorkspace %s is not valid: %s", cw.Name, conditions.GetMessage(cw, tenancyv1alpha1.WorkspaceURLValid))
+					} else if !conditions.IsTrue(cw, tenancyv1alpha1.WorkspaceShardValid) {
+						return false, fmt.Errorf("ClusterWorkspace %s is not valid: %s", cw.Name, conditions.GetMessage(cw, tenancyv1alpha1.WorkspaceShardValid))
```
Comment (aojea) [inline test/e2e/virtual/workspaces/virtual_workspace_test.go:186]: @sttts why is this a hard error, should not retry?
seens a failure here because of this https://github.com/kcp-dev/kcp/runs/5467154069?check_suite_focus=true
